projects/gateway2/krtcollections/policy.go

```diff
@@ -0,0 +1,717 @@
+					FromNs:     string(from.Namespace),
+				})
+			}
+		}
+		return ret
+	})
+	return &RefGrantIndex{refgrants: refgrants, refGrantIndex: refGrantIndex}
+}
+
+func (r *RefGrantIndex) ReferenceAllowed(kctx krt.HandlerContext, fromgk schema.GroupKind, fromns string, to ir.ObjectSource) bool {
+	key := refGrantIndexKey{
+		RefGrantNs: to.Namespace,
+		ToGK:       schema.GroupKind{Group: to.Group, Kind: to.Kind},
+		FromGK:     fromgk,
+		FromNs:     fromns,
+	}
+	if krt.Fetch(kctx, r.refgrants, krt.FilterIndex(r.refGrantIndex, key)) != nil {
+		return true
+	}
+	// try with name:
+	key.ToName = to.Name
+	if krt.Fetch(kctx, r.refgrants, krt.FilterIndex(r.refGrantIndex, key)) != nil {
+		return true
+	}
+	return false
+}
+
+type RouteWrapper struct {
+	Route ir.Route
+}
+
+func (c RouteWrapper) ResourceName() string {
+	os := ir.ObjectSource{
+		Group:     c.Route.GetGroupKind().Group,
+		Kind:      c.Route.GetGroupKind().Kind,
+		Namespace: c.Route.GetNamespace(),
+		Name:      c.Route.GetName(),
+	}
+	return os.ResourceName()
+}
+
+func (c RouteWrapper) Equals(in RouteWrapper) bool {
+	return c.ResourceName() == in.ResourceName() && versionEquals(c.Route.GetSourceObject(), in.Route.GetSourceObject())
+}
+func versionEquals(a, b metav1.Object) bool {
+	var versionEquals bool
+	if a.GetGeneration() != 0 && b.GetGeneration() != 0 {
+		versionEquals = a.GetGeneration() == b.GetGeneration()
+	} else {
+		versionEquals = a.GetResourceVersion() == b.GetResourceVersion()
+	}
+	return versionEquals && a.GetUID() == b.GetUID()
+}
+
+type RoutesIndex struct {
+	routes          krt.Collection[RouteWrapper]
+	httpRoutes      krt.Collection[ir.HttpRouteIR]
+	httpByNamespace krt.Index[string, ir.HttpRouteIR]
+	byTargetRef     krt.Index[types.NamespacedName, RouteWrapper]
+
+	policies  *PolicyIndex
+	refgrants *RefGrantIndex
+	upstreams *UpstreamIndex
+}
+
+func (h *RoutesIndex) HasSynced() bool {
+	return h.httpRoutes.Synced().HasSynced() && h.routes.Synced().HasSynced()
+}
+
+func NewRoutesIndex(krtopts krtutil.KrtOptions, httproutes krt.Collection[*gwv1.HTTPRoute], tcproutes krt.Collection[*gwv1a2.TCPRoute], policies *PolicyIndex, upstreams *UpstreamIndex, refgrants *RefGrantIndex) *RoutesIndex {
+
+	h := &RoutesIndex{policies: policies, refgrants: refgrants, upstreams: upstreams}
+	h.httpRoutes = krt.NewCollection(httproutes, h.transformHttpRoute, krtopts.ToOptions("http-routes-with-policy")...)
+	hr := krt.NewCollection(h.httpRoutes, func(kctx krt.HandlerContext, i ir.HttpRouteIR) *RouteWrapper {
+		return &RouteWrapper{Route: &i}
+	}, krtopts.ToOptions("routes-http-routes-with-policy")...)
+	tr := krt.NewCollection(tcproutes, func(kctx krt.HandlerContext, i *gwv1a2.TCPRoute) *RouteWrapper {
+		t := h.transformTcpRoute(kctx, i)
+		return &RouteWrapper{Route: t}
+	}, krtopts.ToOptions("routes-tcp-routes-with-policy")...)
+	h.routes = krt.JoinCollection([]krt.Collection[RouteWrapper]{hr, tr})
```

Comment:
missed adding our opts and a name here
```suggestion
	h.routes = krt.JoinCollection([]krt.Collection[RouteWrapper]{hr, tr}, krtopts.ToOptions(name: "all-routes-with-policy")...)
```